File: Convex_Hull_Graham.py
import numpy as np
import matplotlib.pyplot as plt
import functools

pivot = None

def orientation(p1,p2,p3):
	val = (p3[1]-p2[1])*(p2[0]-p1[0])-(p2[1]-p1[1])*(p3[0]-p2[0])

	if(val == 0):
		return 0
	if(val>0):
		return 2
	else:
		return 1

def distance(p1,p2):
	return (p1[1]-p2[1])*(p1[1]-p2[1])+(p1[0]-p2[0])*(p1[0]-p2[0])


def PIVOTSORT(p1,p2):
	o = orientation(pivot,p1,p2)
	if(o==0):
		if(distance(pivot,p1) >= distance(pivot,p2)):
			return -1
		else:
			return 1

	if(o==2):
		return 1
	else:
		return -1

def next_to_top(s):
	val = s.pop()
	ret = s.pop()
	s.append(ret)
	s.append(val)
	return ret

def convex_hull(points):
	n = len(points)

	ppo = np.array(points)
	ymin = points[0][1]
	mini = 0

	for i in range(n):
		y = points[i][1]
		if((y<ymin) or ((y==ymin) and points[i][0] < points[mini][0])):
			ymin = y
			mini = i

	points[0],points[mini] = points[mini],points[0]
	global pivot
	pivot = points[0]

	points[1:] = sorted(points[1:],key=functools.cmp_to_key(PIVOTSORT),reverse=True)

	points_new = []

	points_new.append(points[0])
	hullnp = np.array(points_new)
	plt.clf()
	plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
	plt.plot(ppo[:,0],ppo[:,1],".r")
	plt.axis('off')
	plt.show(block=False)
	plt.pause(0.1)
	ii = 1
	while(ii<n):
		while(ii<n-1 and orientation(pivot,points[ii],points[ii+1])==0):
			ii = ii+1
		if(ii<n):
			points_new.append(points[ii])
			hullnp = np.array(points_new)
			plt.clf()
			plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
			plt.plot(ppo[:,0],ppo[:,1],".r")
			plt.axis('off')
			plt.show(block=False)
			plt.pause(0.1)
		ii = ii + 1

	hullnp = np.array(points_new)
	plt.clf()
	plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
	plt.plot(ppo[:,0],ppo[:,1],".r")
	plt.plot([hullnp[-1,0],hullnp[0,0]],[hullnp[-1,1],hullnp[0,1]],'b-',picker=5)
	plt.axis('off')
	plt.show(block=False)
	plt.pause(0.1)
	s = []
	s.append(points_new[0])
	s.append(points_new[1])
	s.append(points_new[2])

	m = len(points_new)

	hullnp = np.array(s)
	plt.clf()
	plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
	plt.plot(ppo[:,0],ppo[:,1],".r")
	plt.axis('off')
	plt.show(block=False)
	plt.pause(0.1)

	for i in range(3,m):
		while(orientation(next_to_top(s),s[-1],points_new[i])!=2):
			s.pop()
		s.append(points_new[i])
		hullnp = np.array(s)
		plt.clf()
		plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
		plt.plot(ppo[:,0],ppo[:,1],".r")
		plt.axis('off')
		plt.show(block=False)
		plt.pause(0.1)
		
	print(s)
	return s


points = [[0,3],[1,1],[2,2],[4,4],[0,0],[1,2],[3,1],[3,3]]

s = convex_hull(points)

print('convex_hull')
ppo = np.array(points)
hullnp = np.array(s)
plt.clf()
plt.plot(hullnp[:,0],hullnp[:,1],'b-',picker=5)
plt.plot(ppo[:,0],ppo[:,1],".r")
plt.plot([hullnp[-1,0],hullnp[0,0]],[hullnp[-1,1],hullnp[0,1]],'b-',picker=5)
plt.axis('off')
plt.show(block=False)
plt.pause(1)
print('Stop')

File: Convex_Hull_Graham.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long int

void swap(pair<ll,ll> &p1,pair<ll,ll> &p2){
    pair<ll,ll> temp = p1;
    p1 = p2;
    p2 = temp;
}

ll orientation(pair<ll,ll> p,pair<ll,ll> q,pair<ll,ll> r){
    ll val = (r.second-q.second)*(q.first-p.first)-
             (q.second-p.second)*(r.first-q.first);
    if(val == 0){
        return 0;
    }
    if(val > 0){
        return 2;
    }else {
        return 1;
    }
}

ll distance(pair<ll,ll> p1,pair<ll,ll> p2){
    return (p1.first-p2.first)*(p1.first-p2.first) +
           (p1.second-p2.second)*(p1.second-p2.second);
}
pair<ll,ll> pivot;

bool PIVOTSORT(pair<ll,ll> p1,pair<ll,ll> p2){
    ll o = orientation(pivot,p1,p2);
    if(o==0){
        if(distance(pivot,p1) >= distance(pivot,p2)){
            return false;
        }else {
            return true;
        }
    }
    if(o==2){
        return true;
    }else {
        return false;
    }
}

pair<ll,ll> next_to_top(stack<pair<ll,ll>> &s){
    pair<ll,ll> val = s.top();
    s.pop();
    pair<ll,ll> ret = s.top();
    s.push(val);
    return ret;
};

void convex_hull(vector<pair<ll,ll>> points){
    ll n = points.size();

    ll ymin = points[0].second;
    ll min = 0;

    for(ll i=0;i<n;i++){
        ll y = points[i].second;
        if((y<ymin)||(y==ymin && points[i].first<points[min].first)){
            ymin = y;
            min = i;
        }
    }

    swap(points[0],points[min]);
    pivot = points[0];
    sort(points.begin()+1,points.end(),PIVOTSORT);

    vector<pair<ll,ll> > points_new;
    points_new.push_back(points[0]);
    for(ll i=1;i<n;i++){
        while(i<n-1 && orientation(pivot,points[i],points[i+1])==0){
            i++;
        }
        if(i<n)
            points_new.push_back(points[i]);
    }

    stack<pair<ll,ll>> s;
    s.push(points_new[0]);
    s.push(points_new[1]);
    s.push(points_new[2]);

    ll m = points_new.size();
    for(ll i=3;i<m;i++){
        while(s.size()>=2 && orientation(next_to_top(s),s.top(),points_new[i])!=2){
            s.pop();
        }
        s.push(points_new[i]);
    }

    while(!s.empty()){
        cout<<"{"<<s.top().first<<","<<s.top().second<<"}\n";
        s.pop();
    }
}

int main(){

    vector<pair<ll,ll> >points;
    points.push_back(make_pair(0,3));
    points.push_back(make_pair(1,1));
    points.push_back(make_pair(2,2));
    points.push_back(make_pair(4,4));
    points.push_back(make_pair(0,0));
    points.push_back(make_pair(1,2));
    points.push_back(make_pair(3,1));
    points.push_back(make_pair(3,3));
    convex_hull(points);
}